File: server/src/services/numberToWords.js
const ONES = [
  "", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine",
  "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen",
  "Seventeen", "Eighteen", "Nineteen"
];
const TENS = [
  "", "", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"
];

function threeDigitsToWords(n) {
  const parts = [];
  if (n >= 100) {
    parts.push(ONES[Math.floor(n / 100)], "Hundred");
    n %= 100;
  }
  if (n >= 20) {
    parts.push(TENS[Math.floor(n / 10)]);
    n %= 10;
    if (n > 0) parts.push(ONES[n]);
  } else if (n > 0) {
    parts.push(ONES[n]);
  }
  return parts.join(" ");
}

// Indian numbering system: crore / lakh / thousand / hundred.
function integerToIndianWords(n) {
  if (n === 0) return "Zero";
  const crore = Math.floor(n / 10000000);
  n %= 10000000;
  const lakh = Math.floor(n / 100000);
  n %= 100000;
  const thousand = Math.floor(n / 1000);
  n %= 1000;
  const hundred = n;

  const segments = [];
  if (crore) segments.push(threeDigitsToWords(crore), "Crore");
  if (lakh) segments.push(threeDigitsToWords(lakh), "Lakh");
  if (thousand) segments.push(threeDigitsToWords(thousand), "Thousand");
  if (hundred) segments.push(threeDigitsToWords(hundred));
  return segments.join(" ").trim();
}

function amountInWordsINR(amount) {
  const rupees = Math.floor(Math.round(amount * 100) / 100);
  const paise = Math.round((amount - rupees) * 100);

  const rupeeWords = integerToIndianWords(rupees);
  let words = `${rupeeWords} Rupee${rupees === 1 ? "" : "s"}`;
  if (paise > 0) {
    words += ` and ${integerToIndianWords(paise)} Paise`;
  }
  return `${words} Only`;
}

module.exports = { amountInWordsINR };

File: go-backend/internal/invoice/numbertowords.go
package invoice

import (
	"fmt"
	"strings"
)

var ones = []string{
	"", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine",
	"Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen",
}

var tens = []string{"", "", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"}

func twoDigitWords(n int) string {
	if n < 20 {
		return ones[n]
	}
	word := tens[n/10]
	if n%10 != 0 {
		word += " " + ones[n%10]
	}
	return word
}

func threeDigitWords(n int) string {
	if n < 100 {
		return twoDigitWords(n)
	}
	word := ones[n/100] + " Hundred"
	if n%100 != 0 {
		word += " " + twoDigitWords(n%100)
	}
	return word
}

// amountInWordsINR renders a rupee amount using the Indian numbering system
// (crore/lakh/thousand), matching the original app's declaration line, e.g.
// "Two Thousand Five Hundred Seventeen Rupees and Twenty Paise Only".
func amountInWordsINR(amount float64) string {
	rupees := int64(amount)
	paise := int(round2((amount-float64(rupees))*100) + 0.0001) // guard against float rounding

	var parts []string
	n := rupees
	if n == 0 {
		parts = append(parts, "Zero")
	} else {
		crore := n / 10000000
		n %= 10000000
		lakh := n / 100000
		n %= 100000
		thousand := n / 1000
		n %= 1000
		remainder := int(n)

		if crore > 0 {
			parts = append(parts, threeDigitWords(int(crore))+" Crore")
		}
		if lakh > 0 {
			parts = append(parts, threeDigitWords(int(lakh))+" Lakh")
		}
		if thousand > 0 {
			parts = append(parts, threeDigitWords(int(thousand))+" Thousand")
		}
		if remainder > 0 {
			parts = append(parts, threeDigitWords(remainder))
		}
	}

	words := strings.Join(parts, " ") + " Rupees"
	if paise > 0 {
		words += fmt.Sprintf(" and %s Paise", threeDigitWords(paise))
	}
	return words + " Only"
}
